### ISSUE TITLE
Shader-based render engine

### ISSUE THREAD

**Issue by dacap** [aseprite/aseprite#3516]
A planned effort we have (as an item of #960) is to rewrite the rendering engine to use shaders (SkSL language, skia shaders). This should fix several issues at the same time and can be a progressive effort:

- [ ] use SkShaders to use CPU or GPU accelerated rendering (similar as we've done with color selectors in #2900)
- [ ] cache rendered tiles (portion of the canvas in an editor/frame)
- [ ] rendering layer groups separately #3225
- [ ] mask layers https://github.com/aseprite/aseprite/issues/459#issuecomment-260010136

I have a minimal work-in-progress commit to start the refactor, but nothing serious yet.

**Comment by dacap** [comment]
We're going to talk with @ckaiser weekly to know how to rewrite the rendering engine and write down the progress here. The idea is to improve the general performance of the editor to edit big sprites/canvases + adding the core for new future features (composite groups, mask layers, fx layers, flip/rotate viewport, etc.).

**Comment by Jonathan-Westfall-Shatter-Fantasy** [comment]
For the new render engine is there a design document that is public on how some of the stuff is planning to be implemented or how the foundational render changes will be carried out. And sorry if this might not be the best place to ask. Wasn't sure where would be best. Here, the forums, or the Skia repo.

Example of information I was curious about:
Is there a specific Skia Graphics milestone release you are aiming to build on? Like Milestone 150 for example.

For the rendering are you planning a type of composite like system. Think like a Render Graph approach where you have a set of shader passes done one after the other that build on the previous checks Render pass. You have the base pixel art that is rendered for the first layer. Than you do passes for blend modes, even a light source, shaders, fx layering, and allowing you to choose when to inject each Render Pass allowing to customize the order of operations.

Example screen shot of Unity's Version of a Render Graph showing each Render Pass and what textures/render targets they touch.
<img width="967" height="433" alt="Image" src="https://github.com/user-attachments/assets/00f1b529-73e1-4edc-abb0-cf28a0ed8b15" />

**Comment by dacap** [comment]
Hi @crowhound, thanks for your interest about this. We weren't able to advance too much on this field.

At first the idea was to implement parts of the new renderer in [ShaderRenderer](https://github.com/aseprite/aseprite/blob/e7debd2a4f6e137534c0bb82bab8e07f9f025d87/src/app/render/shader_renderer.h#L32), but there wasn't too much progress. But the general idea was to replace everything with Skia shaders + caching as much as possible.

The issue is that we need something that can match our cases to render the BrushPreview in real-time between two set of layers (above / below). And the brush preview uses the whole logic of inks (e.g. shading inks), and that is implemented in CPU.

The best case would be to re-implement everything with Skia shaders (brush shapes + inks) and having the chance to use the brush preview just like another layer/node/step in the middle of a render graph/pipeline. But we need a new render architecture for this.

About Skia, m148 seems like the next candidate: https://github.com/aseprite/skia/releases/tag/m148-a29c8d23be (but we can go to m150 if you need something from that version, the [skia-m148](https://github.com/dacap/aseprite/tree/skia-m148) branch is not yet merged in [beta](https://github.com/aseprite/aseprite/tree/beta)).

**Comment by Jonathan-Westfall-Shatter-Fantasy** [comment]
Hello @dacap, thank you for the response. Just want to check is there a set direction planned out for the update to rendering or is it still trying to figure out ideas and how to get around certain limitations while building out the design of the entire renderer architecture. Example the Brush Preview issue you mentioned.

Reason I ask I was thinking of digging more into the Aseprite renderer on my free time and see what all are the moving pieces. 
But wasn't sure how would be the best way to go about it or if their was a  public design document to follow for it.
I am not 100% updated on the exact state of the current aseprite renderer, so was going to see what has changed since last time I looked at it. 

Note the following is just a random example of ideas for certain things I was wondering about how they were going to be done.
Has the idea for how layers and compositing different effects been chosen yet.

Since the idea is to use SkSL shaders for the render pipeline to allow certain customizable functionality for the start of the change and make adding features in the future easier, do you plan to allow each SkShader act as a container for runtime effects allowing them to be layered on each other. Basically using the SKShader as a tree of objects with a using SkRuntimeEffects to support SkShaders::Blend and other stuff.

Example for the first runtime effect would the general Pixel Render Shader. The shader that renders what was actually drawn by artist in the layer. Say an artist wants to do a outline effect on it. Make that the second SkRuntimeEffect. And basically make any FX, custom shader just a chain of SKRuntimeEffects allowing people to layer effects on top of one another to allow compositing a final image.

This idea would also let users add their own custom SkSL shaders. Basically a user could create a new shader effect and when they want to have it applied as one of the children in the SkRuntimeEffect::ChildPtr we just add there set filters to the children[] before calling the SkPaint command to set the shader and draw it onto the canvas as a final render.

The following is just pseudo code, so logically it will be off syntax wise due to skipping the typing of boiler plate code.
Also didn't type out the sksl char* passed into the SkRuntimeEffect::MakeForShader call.

```

void draw(SkCanvas* canvas) 
{


  SkRuntimeEffect::ChildPtr children[] = 
  {
     /*input_1=*/ basePixelShader,   
     /*input_2=*/ outlineShader,
     /*input_3=*/ anyOtherEffects
   };

   // Create an SkShader from our SkSL, with `children` bound to the inputs:

  auto [effect, err] = SkRuntimeEffect::MakeForShader(SkString(sksl));
  sk_sp<SkShader> myShader = effect->makeShader(/*uniforms=*/ nullptr,
                                                /*children=*/ children);

   // Take the SkShader made by the list of children inputs of SkRuntimeEffect 
   // and use it to draw the final rendered image onto the canvas.

   SkPaint p;
   p.setShader(myShader);
   canvas->drawPaint(p);
}
```

**Comment by dacap** [comment]
Yes, the idea is to use SkSl shaders for each step/layer/FX layer/etc. in the future (the list of shaders would be generated using new layer types, wip in #5683).

As I said, the ShaderRenderer is the best we've done so far. We already have shaders to convert indexed -> rgb ( https://github.com/aseprite/aseprite/blob/e7debd2a4f6e137534c0bb82bab8e07f9f025d87/src/app/render/shader_renderer.cpp#L40 ), and other logic implemented, but still a lot to be done.
